### PR TITLE
MARKET-1644 Inner Image Zoom

### PR DESCRIPTION
**Pull request by Fialkaksu**
The InnerImageZoom component allows the user to apply an image zoom effect inside that image in the application.

[MARKET-1644](https://bugs.backendless.com/browse/MARKET-1644)
[UI-Builder](https://develop.backendless.com/app/TestComponentsApp/ui-builder/InnerImageZoom/pages/landing)
[Desktop Preview](https://populardowntown.backendless.app/api/files/ui-builder/containers/InnerImageZoom/index.html?page=landing)

image
onHover:
image

### CODE REVIEW

File: components/bl-innerImageZoom-component/src/index.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    return {
+      backgroundImage: `url(${src})`,
+      backgroundPosition: state,
+    };
+  }, [src, state]);
+
+  const handleMouseMove = (event) => {
+    const { left, top, width, height } = event.target.getBoundingClientRect();
+    const x = ((event.pageX - left - window.pageXOffset) / width) * 100;
+    const y = ((event.pageY - top - window.pageYOffset) / height) * 100;
```
Comment (serhmel) [inline components/bl-innerImageZoom-component/src/index.js:25]: empty line after variables

File: components/bl-innerImageZoom-component/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+| Property                   | Type   | Default value                                                                                                                        | Logic              | Data Binding | UI Setting | Description                         |
+|----------------------------|--------|--------------------------------------------------------------------------------------------------------------------------------------|--------------------|--------------|------------|-------------------------------------|
+| Image Source <br> `source` | *Text* | `https://content.r9cdn.net/rimg/dimg/6d/56/5bd0abd4-city-26896-1670df92320.jpg?width=1366&height=768&xhint=1567&yhint=925&crop=true` | Image Source Logic | YES          | YES        | Specifies the image source to zoom. |
+
+## Events
+
+| Name                | Triggers                                                           | Context Blocks    |
+|---------------------|--------------------------------------------------------------------|-------------------|
+| On Mouse Move Event | Triggered when the user moves the mouse over or touches the image. | `{event: Object}` |
```
Comment (serhmel) [inline components/bl-innerImageZoom-component/README.md:23]: `{event: Object}` => Event: `Object` ?

File: components/bl-innerImageZoom-component/src/index.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  const src = useMemo(() => source, [source]);
+  const zoom = useMemo(() => Math.max(1, zoomIn), [zoomIn]);
+
+  const [zoomPosition, setZoomPosition] = useState('0% 0%');
+
+  const figureStyle = useMemo(() => ({
+    backgroundImage: `url(${src})`,
+    backgroundPosition: zoomPosition,
+    backgroundSize: `${zoom * 100}% ${zoom * 100}%`,
+  }), [src, zoomPosition, zoom]);
```
Comment (serhmel) [inline components/bl-innerImageZoom-component/src/index.js:18]:
```suggestion
  const figureStyle = useMemo(() => ({
    backgroundImage   : `url(${ src })`,
    backgroundPosition: zoomPosition,
    backgroundSize    : `${ zoom * 100 }% ${ zoom * 100 }%`,
  }), [src, zoomPosition, zoom]);
```

File: components/bl-innerImageZoom-component/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+## Configuration
+
+Configuration can be done in the UI Builder or using the Codeless Logic. You have to specify the source of the image that will be displayed in the component and to which the zoom effect will be applied.
+
+## Properties
+
+| Property                   | Type     | Default value                                                                                                                        | Logic              | Data Binding | UI Setting | Description                           |
+|----------------------------|----------|--------------------------------------------------------------------------------------------------------------------------------------|--------------------|--------------|------------|---------------------------------------|
+| Image Source <br> `source` | *Text*   | `https://content.r9cdn.net/rimg/dimg/6d/56/5bd0abd4-city-26896-1670df92320.jpg?width=1366&height=768&xhint=1567&yhint=925&crop=true` | Image Source Logic | YES          | YES        | Specifies the image source to zoom.   |
```
Comment (v-excelsior) [inline components/bl-innerImageZoom-component/README.md:17]: Image Source -> URL

Comment (v-excelsior) [inline components/bl-innerImageZoom-component/README.md:17]: Try to not use external resources for images. Try to use base64 or selfhosted resources.
Also we must discuss it on meeting

File: components/bl-innerImageZoom-component/README.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+| Zoom <br> `zoomIn`         | *Number* | 2                                                                                                                                    |                    | NO           | YES        | Controlled zoom to enlarge the image. |
```
Comment (v-excelsior) [inline components/bl-innerImageZoom-component/README.md:18]: Would be better to use percents for that. 140 looks for me better than 1.4(or 1,4)

File: components/bl-innerImageZoom-component/src/index.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+export default function InnerImageZoomComponent({ component, elRef, eventHandlers }) {
+  const { classList, style, display, source, zoomIn } = component;
+  const { onMouseOver, onMouseOut, onMouseMove } = eventHandlers;
+
+  const src = useMemo(() => source, [source]);
+  const zoom = useMemo(() => Math.max(1, zoomIn), [zoomIn]);
+
+  const [zoomPosition, setZoomPosition] = useState('0% 0%');
+
+  const figureStyle = useMemo(() => ({
```
Comment (v-excelsior) [inline components/bl-innerImageZoom-component/src/index.js:14]: useStyles hook

File: components/bl-innerImageZoom-component/src/index.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
+import { useMemo, useState } from 'react';
+
+const { cn } = BackendlessUI.CSSUtils;
+
+export default function InnerImageZoomComponent({ component, elRef, eventHandlers }) {
+  const { classList, style, display, source, zoomIn } = component;
+  const { onMouseOver, onMouseOut, onMouseMove } = eventHandlers;
+
+  const src = useMemo(() => source, [source]);
```
Comment (v-excelsior) [inline components/bl-innerImageZoom-component/src/index.js:9]: That do nothing

File: components/bl-innerImageZoom-component/src/index.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  const zoom = useMemo(() => Math.max(1, zoomIn), [zoomIn]);
```
Comment (v-excelsior) [inline components/bl-innerImageZoom-component/src/index.js:10]: useMemo calculation is harder than comparison of two numbers

File: components/bl-innerImageZoom-component/src/index.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+  const [zoomPosition, setZoomPosition] = useState('0% 0%');
```
Comment (v-excelsior) [inline components/bl-innerImageZoom-component/src/index.js:12]: '0% 0%' - magic string, try avoid them

const DEFAULT_ZOOM_POSITION

File: components/bl-innerImageZoom-component/src/index.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+const { cn } = BackendlessUI.CSSUtils;
+const DEFAULT_ZOOM_POSITION = '0% 0%';
+
+export default function InnerImageZoomComponent({ component, elRef, eventHandlers }) {
+  const { classList, style, display, source, zoomIn } = component;
+  const { onMouseOver, onMouseOut, onMouseMove } = eventHandlers;
+
+  const zoom = Math.max(100, zoomIn);
+  const [zoomPosition, setZoomPosition] = useState(DEFAULT_ZOOM_POSITION);
+  const figureStyle = useStyles(source, zoomPosition, zoom);
```
Comment (v-excelsior) [inline components/bl-innerImageZoom-component/src/index.js:12]: There are no strict rules, but try to split it in some order.

For example here we could

```
const { classList, style, display, source, zoomIn } = component; //Works with props
const { onMouseOver, onMouseOut, onMouseMove } = eventHandlers;

const [zoomPosition, setZoomPosition] = useState(DEFAULT_ZOOM_POSITION);//Default hooks like useState useRef useMemo

const figureStyle = useStyles(source, zoomPosition, zoom); // Our custom hooks

const zoom = Math.max(100, zoomIn); //regular calculations
```

**Review by v-excelsior — CHANGES_REQUESTED** [review]

**Pull request merged**